### PR TITLE
Fix Anaconda forces payload restart when network (not)change

### PR DESCRIPTION
**Pull request by jkonecny12**
This problem is solved by downloading content of `repomd.xml` file and comparing its MD5 hashes with downloaded `repomd.xml` file from the old repo url.
    
MD5 hash is used here because it takes less memory and we don't need `repomd.xml` content for anything else.
    
The `postSetup()` method was added to the payload. This method can be overridden by `Payload` child classes which needs to do additional tasks on the end of the `restartThread` call from the `PayloadManager` class.
    
*Resolves: rhbz#1373449*
*Resolves: rhbz#1358778*

Second commit is to make docstrings consistent in `Payload` and `YumPayload`.

I want to backport this to Rawhide later.

### CODE REVIEW

File: pyanaconda/packaging/__init__.py [hunk 1]
```diff
@@ -307,6 +306,19 @@ def disableRepo(self, repo_id):
...
+        """Verify availability of existing repositories.
+
+        This method will test if repositories can be reached by URL from
+        actual repositories. It is useful when network settings is changed
+        so that we can verify if repositories are still reachable.
+
+        This method should be overriden.
+        """
+        log.debug("Install method %s is not able to verify availability",
+                  self.data.method.method)
```
Comment (vpodzime) [inline pyanaconda/packaging/__init__.py]: I think this should log the name of the class instead.

File: pyanaconda/packaging/__init__.py [hunk 2]
```diff
@@ -1382,11 +1402,16 @@ def _runThread(self, storage, ksdata, payload, instClass, fallback, checkmount):
...
+        # run payload specific post configuration tasks
+        payload.postSetup()
+
         self._setState(self.STATE_FINISHED)
 
 # Initialize the PayloadManager instance
 payloadMgr = PayloadManager()
 
+
+
```
Comment (vpodzime) [inline pyanaconda/packaging/__init__.py]: Why three blank lines here?

File: pyanaconda/packaging/yumpayload.py [hunk 3]
```diff
@@ -1611,3 +1637,69 @@ def postInstall(self):
...
+
+    def verifyRepoMD(self):
+        """Download and compare with stored repomd.xml file."""
+        new_repomd = self._downloadRepoMD(self._method)
+        new_repomd_hash = self._calculateHash(new_repomd)
+        return new_repomd_hash == self._repomd_hash
+
+    def _calculateHash(self, message):
+        m = hashlib.md5()
+        m.update(message)
```
Comment (vpodzime) [inline pyanaconda/packaging/yumpayload.py]: Why ``message``?

File: pyanaconda/packaging/__init__.py [hunk 4]
```diff
@@ -119,10 +119,9 @@ class PayloadInstallError(PayloadError):
     pass
 
 class Payload(object):
-    """ Payload is an abstract class for OS install delivery methods. """
+    """Payload is an abstract class for OS install delivery methods."""
```
Comment (vpodzime) [inline pyanaconda/packaging/__init__.py]: One comment for this whole commit: tl;dr :)

File: tests/pyanaconda_tests/payload_test.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        """Test verification method."""
+        r = RepoMDMetaHash(DummyPayload(), self._dummyRepo)
+        r.storeRepoMDHash()
+
+        # test if repomd comparision works properly
+        self.assertTrue(r.verifyRepoMD())
+
+        # test if repomd change will be detected
+        with open(self._md_file, 'a') as f:
+            f.seek(os.SEEK_END)
```
Comment (vpodzime) [inline tests/pyanaconda_tests/payload_test.py]: Is this really necessary? I'd expect the ``'a'`` mode to take care of this.

File: pyanaconda/packaging/yumpayload.py (hunk 3 above)
Comment (jkonecny12) [inline pyanaconda/packaging/yumpayload.py]: First I used `input` here but then pylint complain about redefining `input`... so I changed it to a neutral `message`. Do you have better name for this?

File: pyanaconda/packaging/__init__.py (hunk 4 above)
Comment (jkonecny12) [inline pyanaconda/packaging/__init__.py]: I totally understand it :)

File: pyanaconda/packaging/__init__.py (hunk 2 above)
Comment (jkonecny12) [inline pyanaconda/packaging/__init__.py]: Mistake, sorry about that.

File: pyanaconda/packaging/__init__.py (hunk 1 above)
Comment (jkonecny12) [inline pyanaconda/packaging/__init__.py]: I guess it gives more sense. I'll fix that.

File: tests/pyanaconda_tests/payload_test.py (hunk 5 above)
Comment (jkonecny12) [inline tests/pyanaconda_tests/payload_test.py]: Based on the test, it looks like I can remove the `f.seek()` method from here. Thanks for pointing me on this.

**Comment by jkonecny12** [comment]
I just wonder how hard it will be to port this to master/DNF.

File: pyanaconda/packaging/yumpayload.py (hunk 3 above)
Comment (vpodzime) [inline pyanaconda/packaging/yumpayload.py]: ``data``?

Comment (jkonecny12) [inline pyanaconda/packaging/yumpayload.py]: yes, why not :)

**Pull request merged**